alf/batch_norm.py

```diff
@@ -0,0 +1,330 @@
+                self._running_means[i].zero_()
+                self._running_vars[i].fill_(1)
+                self._num_batches_tracked[i].zero_()
+        if self._affine:
+            nn.init.ones_(self._weight)
+            nn.init.zeros_(self._bias)
+
+    def forward(self, input: torch.Tensor):
+        self._check_input_dim(input)
+
+        if self.training or not self._track_running_stats:
+            if self._track_running_stats:
+                current_step = self._current_step
+                if isinstance(current_step,
+                              torch.Tensor) and current_step.ndim != 0:
+                    assert torch.all(current_step == current_step[0]), (
+                        "all current_steps must be same for training.")
+                    current_step = current_step[0]
+                current_step = int(current_step)
+                running_mean = self._running_means[current_step]
+                running_var = self._running_vars[current_step]
+                if not self._clamped:
+                    num_batches_tracked = self._num_batches_tracked[
+                        current_step]
+                    num_batches_tracked.add_(1)
+                    if self._momentum is None:  # use cumulative moving average
+                        exponential_average_factor = 1.0 / float(
+                            num_batches_tracked)
+                    else:  # use exponential moving average
+                        exponential_average_factor = self._momentum
+                else:
+                    exponential_average_factor = 0.0
+            else:
+                running_mean = None
+                running_var = None
+                exponential_average_factor = 0.0
+            return F.batch_norm(
+                input,
+                running_mean,
+                running_var,
+                self._weight,
+                self._bias,
+                # whether the mini-batch stats should be used for normalization
+                # rather than the running stats.
+                # If current_step is out of limit, we will use the running stats
+                # for max_steps - 1 to normalize the batch so that we can keep
+                # training and eval consistent.
+                not self._clamped,
+                exponential_average_factor,
+                self._eps)
+        else:
+            running_means = torch.stack(
+                self._running_means, dim=0)[self._current_step]
+            running_vars = torch.stack(
+                self._running_vars, dim=0)[self._current_step]
+            k = input.ndim - 2
+            if k > 0:
+                ones = [1] * k
+                running_means = running_means.reshape(*running_means.shape,
+                                                      *ones)
+                running_vars = running_vars.reshape(*running_vars.shape, *ones)
+            y = (input - running_means) * (running_vars + self._eps).rsqrt()
```

Comment:
This is a bit different from my understanding of `normalization` - is this meant to be `/ (running_vars + self._eps)` insteadof `*`?

Reply:
rsqrt is reciprocal of sqrt.